### PR TITLE
Fix blocks not starting on new lines (#411)

### PR DESCRIPTION
**Pull request by zepinglee**
This PR modifies the `@` marker regex so that it doesn't necessarily appear at a new line. It also passes all the tests.

### CODE REVIEW

File: tests/splitter_tests/test_splitter_basic.py
Comment (MiWeiss) [inline tests/splitter_tests/test_splitter_basic.py:374]: Could you add a check asserting that the `entry.start_line` values are correct?

**Review by MiWeiss — CHANGES_REQUESTED** [review]
Hi. Thanks a lot for opening the PR. 

I left one minor comment regarding the test above.

A more significant problem is on [line 175](https://github.com/sciunto-org/python-bibtexparser/pull/416/files#diff-9928b8e2ef0ad92b3a5aa9b829f3fb50032950bc14afd79a085b47df6df2005dL175). Currently, we discover if a new block starts with a previous one still open, which typically highlights an error in the parsed file. In such a case, we abort processing the still open block, in deliberate deviation from "strict bibtex parsing". As we only considered `@....` on the start of a new line, the risk of flagging a "false positive" was negligible, but at the same time the user experience increased drastically.

With your change, we would abort the parsing of a block when facing an unexpected `@....` anywhere, e.g, the following could not be parsed:

```
@entry{key, title="all blocks must start with @entry{"} 
```

To reduce this risk, I would love to keep the previous behavior and only abort the parsing of a block if a new block starts on a new line.  Does that make sense?

*Note:*
Another problem, but one which does not have to be fixed is that biber-style comment won't be working with this PR anymore, e.g. the following will likely result in an error.,

```
% @entry{key, 
% title="all blocks must start with @entry{"} 
```
However, we technically still don't support biber comments and a clean solution is required there anyways, thus I'd leave this to #372

File: tests/splitter_tests/test_splitter_basic.py
Comment (zepinglee) [reply tests/splitter_tests/test_splitter_basic.py:374]: Sure.

**Comment by zepinglee** [comment]
> A more significant problem is on [line 175](https://github.com/sciunto-org/python-bibtexparser/pull/416/files#diff-9928b8e2ef0ad92b3a5aa9b829f3fb50032950bc14afd79a085b47df6df2005dL175). Currently, we discover if a new block starts with a previous one still open, which typically highlights an error in the parsed file. In such a case, we abort processing the still open block, in deliberate deviation from "strict bibtex parsing". As we only considered `@....` on the start of a new line, the risk of flagging a "false positive" was negligible, but at the same time the user experience increased drastically.

I don't think the `_move_to_comma_or_closing_curly_bracket()` method where line 175 is located matches the original BibTeX's behavior. BibTeX only accepts strictly brace-balanced string in a literal field token (in terms of the `bibtex.web`) even when the outer delimiters are quotes. The braces or quotation marks in a field token are not treated as escaped characters. In addition, an `@` character doesn't break the field.

https://github.com/sciunto-org/python-bibtexparser/blob/770eb001349dcad48149bc101f07cfc88282d8b3/bibtexparser/splitter.py#L131-L189

In the example of `@entry{key, title="all blocks must start with @entry{"}`, the second quotation mark is in brace level 1 and it doesn't terminate the field token. The final right brace make the string balanced and BibTeX look for another quotation mark to end the field. This is verified by running BibTeX and the raised error `Illegal end of database file`. Yet `@entry{key, title="all blocks must start with @entry{"}"}` works fine with BibTeX and the `title` field is `all blocks must start with @entry{"}`.

I'm planning to rewrite this method and make another PR.

File: tests/splitter_tests/test_splitter_basic.py
Comment (zepinglee) [reply tests/splitter_tests/test_splitter_basic.py]: I've added these assertions in d9a3245. Also note that I removed fields other than `author`, `title`, and `doi` to make less lines.

**Comment by MiWeiss** [comment]
Thanks for your PR update and for your long comment.

#### First things first:
Your argument convinced me that a plain `@entry{` would be very rare, and I thus see no reason not to merge this PR anymore. We may change on this feature in the future, based on received feedback, but for now I guess merging this PR is a clear improvement of things!

#### Regarding the larger argument you made:
 This library is deliberately not 100% consistent with bibtex, but it attempts to be somewhat more fault-tolerant: We aim to be able to handle cases when there are small mistakes in the bibtex file gently, without messing up the parsing of subsequent blocks, or even prevent parsing at all. This is a request we received numerous times reagarding bibtexparser v1. Take the following example:

```
@string{myKey = {myValue))

@string{anotherString = {anotherValue}}
```

While the intention of the writer here is pretty clear, bibtex would not be happy: Note the `))` on the first line which should be `}}`. Technically, the braces are not balanced after the first line, thus not allowing for a new string on the third line. Previously, in `bibtexparser v2`, we thus aborted the parsing of a block if a line started with a "block start", e.g. a `@entry{`. This allowed a "fresh" start with that block. After merging this PR, we now do this even if the "block start" appears anywhere, not just on a new line. 

Thus, I am very happy to review any improvement you make to the parsing (and I am sure there are plenty to be made ;-)), but please keep in mind that we take a deliberate trade-off between sticking to the pure bibtex rules, and user friendly parsing of imperfect files. If in doubt, before investing too much time, feel free to ask for mine (and anyones) comment in a seperate issue before investing too much time.

#### Most importantly
 Thanks again for your contribution, and I am looking forward to your next PRs.

**Review by MiWeiss — APPROVED** [review]

**Pull request merged**
